sdk/testing/stepwise/stepwise.go

```diff
@@ -0,0 +1,332 @@
+	// "uuid":
+	//
+	// vault-plugin-secrets-uuid/
+	// - backend.go
+	// - cmd/
+	// ----uuid/
+	// ------main.go
+	// - path_generate.go
+	//
+	PluginName string
+}
+
+// Step represents a single step of a test Case
+type Step struct {
+	// Operation defines what action is being taken in this step; write, read,
+	// delete, et. al.
+	Operation Operation
+
+	// Path is the localized request path. The mount prefix, namespace, and
+	// optionally "auth" will be automatically added.
+	Path string
+
+	// Arguments to pass in the request. These arguments represent payloads sent
+	// to the API.
+	Data map[string]interface{}
+
+	// Assert is a function that is called after this step is executed in order to
+	// test that the step executed successfully. If this is not set, then the next
+	// step will be called
+	Assert AssertionFunc
+
+	// Unauthenticated will make the request unauthenticated.
+	Unauthenticated bool
+}
+
+// AssertionFunc is the callback used for Assert in Steps.
+type AssertionFunc func(*api.Secret, error) error
+
+// Case represents a scenario we want to test which involves a series of
+// steps to be followed sequentially, evaluating the results after each step.
+type Case struct {
+	// Environment is used to setup the Vault instance and provide the client that
+	// will be used to drive the tests
+	Environment Environment
+
+	// Steps are the set of operations that are run for this test case. During
+	// execution each step will be logged to output with a 1-based index as it is
+	// ran, with the first step logged as step '1' and not step '0'.
+	Steps []Step
+
+	// SkipTeardown allows the Environment TeardownFunc to be skipped, leaving any
+	// infrastructure created after the test exists. Depending on the Environment
+	// used this could incur costs the user is responsible for.
+	SkipTeardown bool
+}
+
+// Run performs an acceptance test on a backend with the given test case.
+//
+// Tests are not run unless an environmental variable "VAULT_ACC" is
+// set to some non-empty value. This is to avoid test cases surprising
+// a user by creating real resources.
+//
+// Tests will fail unless the verbose flag (`go test -v`, or explicitly
+// the "-test.v" flag) is set. Because some acceptance tests take quite
+// long, we require the verbose flag so users are able to see progress
+// output.
+func Run(tt TestT, c Case) {
+	tt.Helper()
+	// We only run acceptance tests if an env var is set because they're
+	// slow and generally require some outside configuration.
+	checkShouldRun(tt)
+
+	if c.Environment == nil {
+		tt.Fatal("nil driver in acceptance test")
+		// return here only used during testing when using mockT type, otherwise
+		// Fatal will exit
+		return
+	}
+
+	logger := logging.NewVaultLogger(log.Trace)
+
+	if err := c.Environment.Setup(); err != nil {
+		driverErr := fmt.Errorf("error setting up driver: %w", err)
+		if !c.SkipTeardown {
+			if err := c.Environment.Teardown(); err != nil {
```

Comment:
Is calling `Teardown` necessary here if the `Setup` failed in the first place?

Reply:
Likely not; I changed that in 483bc3080 to simply call `Fatal` with the error.